Netplan status --diff

daniloegea opened the pull request:
## Description

This PR introduces `netplan status --diff`. It will use the `state_diff` module to output the inconsistencies in the network configuration between the system and the netplan configuration.

The diff will be displayed the same way `netplan status` already displays information. Even though it's a nice approach, plugging the `diff` display code in the existing `status` code made things a bit more complicated. I have refactored it (only after have plugged the code, sorry for that) and split the code responsible for each block of information in their own methods.

It also adds the information about bonds, bridges and VRFs introduced by #420.

## Checklist

- [ ] Runs `make check` successfully.
- [ ] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

slyon commented on netplan_cli/cli/commands/status.py:
```diff
@@ -88,18 +88,164 @@ def pretty_print(self, data: JSON, total: int, _console_width=None) -> None:
+                return diff.get('system_state', {}).get('missing_dhcp6_address', False)
+        return False
+
+    def _get_missing_system_bond_link(self, ifname) -> str:
+        if self.state_diff:
+            if diff := self.state_diff['interfaces'].get(ifname):
+                return diff.get('system_state', {}).get('missing_bond_link', '')
+        return ''
+
+    def _get_missing_netplan_bond_link(self, ifname) -> str:
+        if self.state_diff:
+            if diff := self.state_diff['interfaces'].get(ifname):
+                return diff.get('netplan_state', {}).get('missing_bond_link', '')
+        return ''
```
nitpick (non-blocking): This (and surrounding) methods feel like lots of duplicated code, doing almost the same thing for `system_state` and `netplan_state`, I wonder if we could pass the relevant `diff.get('NETPLAN/SYSTEM_state')` data as an input variable, to use the same method for both cases?

slyon commented on netplan_cli/cli/commands/status.py:
```diff
@@ -88,18 +88,164 @@ def pretty_print(self, data: JSON, total: int, _console_width=None) -> None:
+                return diff.get('netplan_state', {}).get('missing_bond_link', '')
+        return ''
+
+    def _get_missing_system_bridge_link(self, ifname) -> str:
+        if self.state_diff:
+            if diff := self.state_diff['interfaces'].get(ifname):
+                return diff.get('system_state', {}).get('missing_bridge_link', '')
+        return ''
```
question: In `_get_missing_system_macaddress` we're returning `None` for an undefined string, here (and in some other, similar methods) it's the empty string.. Why the difference?

daniloegea replied:
No reason, I don't why I did that :)

slyon commented on netplan_cli/cli/commands/status.py:
```diff
@@ -109,184 +255,613 @@ def pretty_print(self, data: JSON, total: int, _console_width=None) -> None:
+            self.pprint('{} inactive interfaces hidden. Use "--all" to show all.'.format(hidden))
+
+        if (self.diff and not self.diff_only):
+            self.pprint('Use "--diff-only" to omit the information that is consistent between the system and Netplan.')
```
nitpick: This looks a bit odd, due to the `'netplan.str': 'grey62',` definition in the diff Theme. Can we make this line to use the usual theme?

daniloegea replied:
To use the other theme I'd need to create a new Console() so I guess I'll just add [yellow] around the --diff-only. I think it's the only part being affected by the theme...

slyon commented on tests/cli/test_status.py:
```diff
@@ -503,3 +512,119 @@ def test_call_cli_networkd_masked(self, is_masked_mock, is_active_mock):
+        state = SystemConfigState()
+        networkd_mock.return_value = state.process_networkd(NETWORKD)
+        out = self._call(['--diff-only'])
+        self.assertEqual('', out)
```
question (non-blocking): Should we put some message in case where there isn't any diff to show?

slyon replied:
Agreed. That also seems to be what `git diff` does.

slyon commented on tests/cli/test_status_diff.py:
```diff
@@ -0,0 +1,1188 @@
+
+        # Same test for --diff-only
+
+        expected = '''+ ●  1: lo ethernet UNKNOWN/UP (unmanaged)
+
+'''
```
question (non-blocking): Should we trim unused whitespace in the output?

daniloegea replied:
I suppose it would be nice. It will require checking if the interface we are displaying is or is not the last one. I'll do that

slyon commented on tests/cli/test_status_diff.py:
```diff
@@ -0,0 +1,1188 @@
+               Routes: ::1 metric 256
+
+  ●  2: enp5s0 ethernet UP (networkd: enp5s0)
++         MAC Address: 00:16:3e:71:d0:1f (Red Hat, Inc.)
+-                      aa:bb:cc:dd:ee:ff (Red Hat, Inc.)
```
parise: Thanks for adding the +/- signs, so we can test this code even without the color theming.

slyon reviewed (commented):
Thank you very much Danilo, I very much like the looks of this!
And wow! Crafting all the data for `test_status_diff.py` must have been lots of busywork, but it's worth the effort. Thanks for going the extra mile!

Should this PR become a squash merge (due to the refactoring happening only after the previous commits)?

I left some non-critical inline-comments, which I'd like to see discussed before merging it.
Also, I think we should update the manpage/docs around `status --diff[-only]`, but this could (should?) happen in a separate PR, so we can more easily get Robert's feedback, too.

A final (non-blocking) thought: We should consider using `ip addr` information to decide if an IP address got assigned through DHCP, looking at the `dynamic` flag. As DHCPv6 IP addresses are mostly not correctly detected on my system. Probably also in a separate PR, as this one is already pretty big.
```
$ ip -d -j addr show dev lan0 | jq
[...]
    "addr_info": [
      {
        "family": "inet",
        "local": "192.168.178.141",
        "prefixlen": 24,
        [...]
        "scope": "global",
        "dynamic": true,
      },
      {
        "family": "inet6",
        "local": "2001:16b8:bdba:b100:fae4:3bff:fe2d:3bb7",
        "prefixlen": 64,
        [...]
        "scope": "global",
        "dynamic": true,
      }
    ]
```

daniloegea commented on netplan_cli/cli/utils.py:
```diff
@@ -213,6 +213,20 @@ def find_matching_iface(interfaces: list, netdef):
+                split_line = line.split()
+                if len(split_line) == 2 and split_line[0].isnumeric():
+                    lookup_table[int(split_line[0])] = split_line[1]
+    except Exception:
+        return {}
```
Ok, but maybe only in debug mode. It's not really critical and I don't want the message to appear every time you call `netplan status`.

daniloegea commented on netplan_cli/cli/commands/status.py:
```diff
@@ -61,22 +65,18 @@ def run(self):
                                  help='Show extra information')
         self.parser.add_argument('-f', '--format', default='tabular',
                                  help='Output in machine readable `json` or `yaml` format')
+        self.parser.add_argument('--diff', action='store_true',
+                                 help='Show the differences between the system\'s and netplan\'s states')
+        self.parser.add_argument('--diff-only', action='store_true',
```
hmm I don't know, every time I see `-d` I automatically assume it's for `--debug`. Maybe we should avoid that.

daniloegea commented on netplan_cli/cli/commands/status.py:
```diff
@@ -109,184 +255,613 @@ def pretty_print(self, data: JSON, total: int, _console_width=None) -> None:
+
+            self.state_diff = diff_state.get_diff(self.ifname)
+
+            if output_format == 'json':
+                print(json.dumps(self.state_diff, cls=DiffJSONEncoder))
+                return
+            elif output_format == 'yaml':
```
That's a good point. I fixed that for the normal output for missed for json and yaml. I'll look at that.

daniloegea commented:
Thank you, Lukas. I think it's ready for another round of code review. I believe I addressed most of your comments, I think the main one being the linebreak at the end of the output and the code duplication.

Regarding squashing all the commits, I don't know. While there is a little mess in the middle, there are multiple commits that are well separated. Maybe all the ones names as `status` and `status/diff` (and the tests I guess) could be squashed together, but not the rest.

slyon reviewed (approved):
Thank you for the fixes! This is looking really good. I like using the new resulting sub-command a lot!

Could you please rebase and maybe clean up the git history a bit, before we merge it?

Besides that I left have some tiny comments
* Unknown device type: "other": warnings on my system (probably unrelated, but easy to fix)
* `_get_missing_property_str` can return `None` or the empty string
* Should we drop the `MISSING` state, to unclutter the output?
* Should we display unknown interfaces types as `other` instead of `unknown` to stay in line with `status.py`?

I'll push some commits on top of your branch to fix the mentioned issues. Feel free to integrate them into your rebase or revert/drop them with a comment.

daniloegea commented on netplan_cli/cli/commands/status.py:
```diff
@@ -121,113 +121,83 @@ def _is_interface_missing_in_netplan(self, ifname) -> bool:
-                return diff.get('system_state', {}).get('missing_nameservers_addresses', [])
-        return []
+                return diff.get(state, {}).get(property)
+        return ''
```
I think it's better to fix the .get() and return an empty string by default, just for consistency with the return type hint (as None is not a str)

slyon replied:
wfm.

daniloegea commented on netplan_cli/cli/commands/status.py:
```diff
@@ -109,184 +225,625 @@ def pretty_print(self, data: JSON, total: int, _console_width=None) -> None:
+                idx='',
+                name=interface,
+                type=properties.get('type'),
+                state='MISSING',
```
that makes sense, I'll drop it.

slyon reviewed (approved):
Thanks a lot. That works for me!
